keps/sig-storage/5381-mutable-pv-affinity/README.md

```diff
@@ -0,0 +1,872 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes to make `PersistentVolume.spec.nodeAffinity` field mutable,
+making it possible to change the affinity with VolumeAttributeClass.
+This allows user to migrate data or enabling features without
+interrupting workloads.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, `PersistentVolume.spec.nodeAffinity` is set at creation time and cannot be changed.
+But user may modify the volume to
+taking advantage of new features provided by the storage provider,
+or accommodate to the changes of business requirements.
+These modification can be expressed by `VolumeAttributeClass` in Kubernetes.
+But sometimes, A modification to volume comes with change to its accessibility, such as:
+1. migration of data from one zone to regional storage;
+2. enabling features that is not supported by all the client nodes.
+
+In these scenarios, the `nodeAffinity` becomes inaccurate,
+causing the scheduler to make decisions based on outdated information.
+This results in pods:
+* being scheduled to nodes that cannot access the volume, getting stuck in a `ContainerCreating` state;
+* or being rejected from nodes that actually can access the volume, getting stuck in a `Pending` state.
+
+By making `PersistentVolume.spec.nodeAffinity` field mutable,
+we give storage providers a chance to propagate latest accessibility requirement to the scheduler,
+improving the reliability of stateful pod scheduling.
+
+### Goals
+
+- Make `PersistentVolume.spec.nodeAffinity` field mutable.
+- Enable CSI drivers to return a new accessibility requirement on ControllerModifyVolume.
+
+### Non-Goals
+
+- Modifying the core scheduling logic of Kubernetes.
+- Implementing cloud provider-specific solutions within Kubernetes core.
+- Re-scheduling running pods with volumes being modified,
+  or interrupting workloads in any way.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+1. Change APIServer validation to allow `PersistentVolume.spec.nodeAffinity` to be mutable.
+
+2. Change CSI Specification to allow `ControllerModifyVolume` to return a new accessibility requirement.
```

Comment:
Point 2 and 3, though not directly compatible with the title, are connected as upstream and downstream changes. They're currently combined, but if needed, we'll create a new KEP for separation later.